src/computability/regular_expressions.lean

```diff
@@ -0,0 +1,295 @@
+          refine ⟨ t, u, by finish, _, hQ ⟩,
+          rw rmatch at hP,
+          convert hP,
+          finish } } },
+    { rw ih,
+      split;
+      rintro ⟨ t, u, h, hP, hQ ⟩,
+      { exact ⟨ a :: t, u, by tauto ⟩ },
+      { cases t with b t,
+        { contradiction },
+        { refine ⟨ t, u, by finish, _, hQ ⟩,
+          rw rmatch at hP,
+          convert hP,
+          finish } } } }
+end
+
+@[simp] lemma mul_rmatch_iff (P Q : regular_expression α) (x : list α) :
+  (P * Q).rmatch x ↔ ∃ t u : list α, x = t ++ u ∧ P.rmatch t ∧ Q.rmatch u := RComp_rmatch_iff P Q x
+
+lemma RStar_rmatch_iff (P : regular_expression α) : ∀ (x : list α),
+  (RStar P).rmatch x ↔ ∃ S : list (list α), x = S.join ∧ ∀ t ∈ S, ¬(list.empty t) ∧ P.rmatch t
+| x :=
+begin
+  have IH := λ t (h : list.length t < list.length x), RStar_rmatch_iff t,
+  clear RStar_rmatch_iff,
+  split,
+  { cases x with a x,
+    { intro,
+      fconstructor,
+      exact [],
+      tauto },
+    { rw [rmatch, feed, RComp_rmatch_iff],
+      rintro ⟨ t, u, hs, ht, hu ⟩,
+      have hwf : u.length < (list.cons a x).length,
+      { rw [hs, list.length_cons, list.length_append],
+        omega },
+      rw IH _ hwf at hu,
+      rcases hu with ⟨ S', hsum, helem ⟩,
+      use (a :: t) :: S',
+      split,
+      { finish },
+      { intros t' ht',
+        cases ht' with ht' ht',
+        { rw ht',
+          exact ⟨ dec_trivial, ht ⟩ },
+        { exact helem _ ht' } } } },
+  { rintro ⟨ S, hsum, helem ⟩,
+    cases x with a x,
+    { dec_trivial },
+    { rw [rmatch, feed, RComp_rmatch_iff],
+      cases S with t' U,
+      { exact ⟨ [], [], by tauto ⟩ },
+      { cases t' with b t,
+        { finish },
+        refine ⟨ t, U.join, by finish, _, _ ⟩,
+        { specialize helem (b :: t) _,
+          rw rmatch at helem,
+          convert helem.2,
+          finish,
+          finish },
+        { have hwf : U.join.length < (list.cons a x).length,
+          { rw hsum,
+            simp only
+              [list.join, list.length_append, list.cons_append, list.length_join, list.length],
+            omega },
+          rw IH _ hwf,
+          refine ⟨ U, rfl, λ t h, helem t _ ⟩,
+          right,
+          assumption } } } }
+end
+using_well_founded {
+  rel_tac := λ _ _, `[exact ⟨(λ L₁ L₂ : list _, L₁.length < L₂.length), inv_image.wf _ nat.lt_wf⟩]
+}
+
+lemma add_assoc (P Q R : regular_expression α) :
+  (P + Q) + R ≈ P + (Q + R) := by finish
+
+lemma add_comm (P Q : regular_expression α) :
+  P + Q ≈ Q + P := by finish
+
+lemma mul_add (P Q R : regular_expression α) :
+  P * (Q + R) ≈ (P * Q) + (P * R) :=
+begin
+  intro x,
+  simp only [mul_rmatch_iff, add_rmatch_iff],
+  split,
+  { rintro ⟨ s, t, hsum, hP, (hQ | hR) ⟩,
+    left,
+    rotate,
+    right,
+    all_goals
+    { use [s, t],
+      tauto } },
+  { rintro (h | h);
+    rcases h with ⟨ s, t, hsum, hP, h ⟩;
+    use [s, t];
+    tauto }
+end
+
+lemma add_zero (P : regular_expression α) : P + 0 ≈ P := by finish
```

Comment:
These lemmas will be hard to use because `≈` is not equality.

I would suggest to define a language type (as described in the other comment).  You can then prove `P + 0 = P` for `P : language α` (and this is a great simp lemma).  For regexes, you'd just write `P.matches = Q.matches` instead of `P ≈ Q`.